driver/src/main/java/org/neo4j/driver/internal/cluster/LoadBalancer.java

```diff
@@ -0,0 +1,226 @@
+                routers.update( cluster.routers, removed );
+                // purge connections to removed addresses
+                for ( BoltServerAddress address : removed )
+                {
+                    connections.purge( address );
+                }
+            }
+            catch ( InterruptedException e )
+            {
+                throw new ServiceUnavailableException( "Thread was interrupted while establishing connection.", e );
+            }
+        }
+    }
+
+    private ClusterComposition lookupRoutingTable() throws InterruptedException, ServiceUnavailableException
+    {
+        int size = routers.size(), failures = 0;
+        if ( size == 0 )
+        {
+            throw new ServiceUnavailableException( NO_ROUTERS_AVAILABLE );
+        }
+        for ( long start = clock.millis(), delay = 0; ; delay = Math.max( settings.retryTimeoutDelay, delay * 2 ) )
+        {
+            long waitTime = start + delay - clock.millis();
+            if ( waitTime > 0 )
+            {
+                clock.sleep( waitTime );
+            }
+            start = clock.millis();
+            for ( int i = 0; i < size; i++ )
+            {
+                BoltServerAddress address = routers.next();
+                if ( address == null )
+                {
+                    throw new ServiceUnavailableException( NO_ROUTERS_AVAILABLE );
+                }
+                ClusterComposition cluster;
+                try ( Connection connection = connections.acquire( address ) )
+                {
+                    cluster = provider.getClusterComposition( connection );
+                }
+                catch ( Exception e )
+                {
+                    log.error( String.format( "Failed to connect to routing server '%s'.", address ), e );
+                    continue;
+                }
+                if ( cluster == null || !cluster.isValid() )
+                {
+                    log.info(
+                            "Server <%s> unable to perform routing capability, dropping from list of routers.",
+                            address );
+                    routers.remove( address );
+                    if ( --size == 0 )
+                    {
+                        throw new ServiceUnavailableException( NO_ROUTERS_AVAILABLE );
+                    }
+                }
+                else
+                {
+                    return cluster;
+                }
+            }
+            if ( ++failures > settings.maxRoutingFailures )
+            {
+                throw new ServiceUnavailableException( NO_ROUTERS_AVAILABLE );
+            }
+        }
+    }
+
+    private synchronized void forget( BoltServerAddress address )
+    {
+        // first remove from the load balancer, to prevent concurrent threads from making connections to them
+        readers.remove( address );
+        writers.remove( address );
+        routers.remove( address );
```

Comment:
I wonder if removing the address from `routers` really is a good idea.
`forget(...)` is called when a connection fails, and during routing we _don't_ remove the address from the set of routers, perhaps we shouldn't in the general case of connection failure either.